Remove anaconda

ajeddeloh opened the pull request:
* Only for qemu right now. Trying to make sense of the buildextend-metal, it looks like it's running anaconda again?
* Only tested with fcos
* We ought to not use yaml for image.yaml, it's a pain to parse
* Doesn't actually remove anaconda, just switches to not using it

cgwalters commented on src/create_disk.sh:
```diff
@@ -0,0 +1,68 @@
+	-n 1:0:+128M -c 1:boot \
+	-n 2:0:+128M -c 2:EFI-SYSTEM -t 2:C12A7328-F81F-11D2-BA4B-00A0C93EC93B \
+	-n 3:0:+128M -c 3:BIOS-BOOT  -t 3:21686148-6449-6E6F-744E-656564454649 \
+	-n 4:0:0     -c 4:root       -t 4:4F68BCE3-E8CD-4DB1-96E7-FBCAF984B709
```
Random idea - change the order of the partitions on each build to avoid having people rely on it, and instead use the labels.

ajeddeloh replied:
I'm not sure how I feel about though I'm open to it. Regardless, lets punt that to a different PR/discussion?

ajeddeloh replied:
Marking this as resolved unless you really want it in this PR in particular.

cgwalters commented on src/create_disk.sh:
```diff
@@ -0,0 +1,68 @@
+# init the ostree
+ostree admin init-fs rootfs
+ostree pull-local "$ostree" --repo rootfs/ostree/repo
+ostree admin os-init fedora-coreos --sysroot rootfs
```
Let's continue templating this from the manifest.

cgwalters commented:
> We ought to not use yaml for image.yaml, it's a pain to parse

From shell script yes but Python is good for that right?

ajeddeloh commented:
> From shell script yes but Python is good for that right?

Misc idea: re-render to a file in tmp it as json using python. Then bash utils can use `jq` and python can use the native python implementation.

ajeddeloh commented:
Updated so we build bios+uefi images now.

lucab commented on src/create_disk.sh:
```diff
@@ -0,0 +1,77 @@
+ostree admin init-fs rootfs
+ostree pull-local "$ostree" --repo rootfs/ostree/repo
+ostree admin os-init fedora-coreos --sysroot rootfs
+kargs='root=/dev/disk/by-label/root console=ttyS0,115200n8 console=ttyS0 rootflags=defaults,prjquota rw $ignition_firstboot'
```
Can we move the middle part of this to two variables (`$consoleflags` and `$rootflags`) that can be overridden via environment?

cgwalters replied:
Also we need to use `VM_TERMINAL` instead of `ttyS0`.

ajeddeloh replied:
@lucab what environment precisely? This is run in supermin, so the environment is pretty sparse.

The kargs currently live in a kickstart. Seems like they ought to live in `image.yaml`?

cgwalters replied:
https://github.com/coreos/coreos-assembler/pull/525

ajeddeloh replied:
Summarizing some discussion on `#fedora-coreos`: We're going to keep the kargs that relate to this script in this script but move the other kargs to cmd-build since we can do thing like templating (for things like `$VM_TERMINAL` there since it's not in supermin. We may move them later, but that's for another PR.

cgwalters commented:
> Misc idea: re-render to a file in tmp it as json using python. Then bash utils can use jq and python can use the native python implementation.

Sure, fine by me.

ajeddeloh commented:
I'll deal with doing the yaml->json re-rendering later in another PR. Doesn't need to happen in this one.

ajeddeloh commented on src/vmdeps.txt:
```diff
@@ -17,3 +17,5 @@ selinux-policy selinux-policy-targeted policycoreutils
 # coreos-assembler
 #FEDORA python3 python3-gobject-base buildah podman skopeo iptables iptables-libs
+
+gdisk xfsprogs e2fsprogs grub2 dosfstools grub2-efi-x64-modules
```
Do you think this is necessary now or can it be in a follow up?

cgwalters replied:
Let's ensure the right people are pinged here and I think we can get some resources to do basic sanity checking.  Another good approach would be to just look at the partitions from a ppc64 install from Anaconda, etc.

jcajka replied:
@ajeddeloh for ppc64le it should be as mentioned in https://github.com/coreos/fedora-coreos-tracker/issues/78#issuecomment-503571141

ajeddeloh replied:
Talked with some some folks about other arches and this will need conditionals in the future, but for now it's fine. Marking resolved for now and we'll revisit when switching more arches from anaconda.

cgwalters commented on src/cmd-build:
```diff
@@ -242,9 +242,17 @@ img_base=tmp/${imageprefix}-base.qcow2
+build_image() {
+	local size
+	size="$(python3 -c 'import sys, yaml; print(yaml.load(sys.stdin)["size"])' < "$configdir/image.yaml")G"
+	qemu-img create -f qcow2 "$workdir/diskimage.qcow2" "$size"
```
Instead of `$workdir` this should be in the build tmpdir `$(pwd)` (which would require passing it as an argument to `runvm` for cleanliness)

ajeddeloh replied:
In that case might as well pass the base image path directly to runvm, right?

ajeddeloh commented:
Addressed things besides the mutli-arch bit, waiting to learn a bit more about ppc64le on that.

Do we need to support arm64 on non-uefi platforms?

r4f4 commented:
No, aarch64 is EFI-only.

jlebon commented on src/cmd-build:
```diff
@@ -242,9 +242,19 @@ img_base=tmp/${imageprefix}-base.qcow2
+	local size kargs
+	size="$(python3 -c 'import sys, yaml; print(yaml.load(sys.stdin)["size"])' < "$configdir/image.yaml")G"
+	kargs="$(python3 -c 'import sys, yaml; args = yaml.load(sys.stdin)["extra-kargs"]; print(" ".join(args))' < "$configdir/image.yaml")"
+	kargs="$kargs console=${VM_TERMINAL} console=${VM_TERMINAL},115200n8"
```
Hmm, looks like we are specifying the same console arg twice? I think we want something like this instead: `console=tty0 console=${VM_TERMINAL}`, right?

ajeddeloh replied:
I don't know exactly, so I'll defer to you.

jlebon commented on src/create_disk.sh:
```diff
@@ -0,0 +1,82 @@
+#!/bin/sh
```
This is minor, but how about we `set -euo pipefail` right from the start?
Seems like the only change is just doing a `if [ $# -ne 6 ]; then usage; fi` and then you can `shift` in peace.

Especially when debugging stuff like this running inside a VM running inside a container, `set -u` is super useful.

ashcrow commented:
cc @darkmuggle @miabbott @imcleod

ajeddeloh commented:
Added a check so we fall back to anaconda on non-x86_64 arches. Addressed @jlebon's comments.

Do we want to handle buildextend-metal here or in another PR?
How important is it to use the size estimation for that?
Does it make sense to use the size estimation for qemu as well?

cgwalters commented:
> Added a check so we fall back to anaconda on non-x86_64 arches.

As a short term that's probably OK, will make it easier to test without disrupting the multi-arch porting in progress.  

> How important is it to use the size estimation for that?

If we don't we're back in the guessing game of how big the disk should be...

> Does it make sense to use the size estimation for qemu as well?

No, that's where `size` from `image.yaml` comes in.

jlebon commented:
> Do we want to handle buildextend-metal here or in another PR?

I'm assuming `buildextend-metal` still works fine after this, right? If so, then yeah, I don't see the harm in doing it as a follow-up!

ajeddeloh commented:
Right now `cmd-buildextend-metal` recreates entirely new images (it has to, the partition and FS are smaller). so right now the BM images are still created using anaconda.

So it sounds like this should not wait for the buildextend-metal (be and qemu only for the moment) and I'll follow up with the corresponding `cmd-buildextend-metal` changes.

Is `cmd-buildextend-metal` expected to work with non-x86 images? It hardcodes BIOS and UEFI options.

cgwalters commented:
> So it sounds like this should not wait for the buildextend-metal (be and qemu only for the moment) and I'll follow up with the corresponding cmd-buildextend-metal changes.

I'm also OK with that.

ajeddeloh commented:
Ok, this should be ready for more review then. Also gated the changes to `gf-platformid` to only take effect on x86_64 as well.

jlebon commented on src/gf-platformid:
```diff
@@ -41,19 +41,21 @@ coreos_gf_run_mount "${tmp_dest}"
-    grubcfg_path=$(coreos_gf glob-expand /boot/efi/EFI/*/grub.cfg)
-else
-    grubcfg_path=/boot/loader/grub.cfg
+if [ "$arch" -ne "x86_64" ]; then
```
This is called by `buildextend-metal` too though. Needs a `|| [ "$platformid" == "metal" ]`, right?

ajeddeloh replied:
Oh yup, good catch. Fixed.

jlebon commented:
This looks good to me! (Though we can probably squash 7ef9087 into ef0450c, right?)
Let's get https://github.com/coreos/fedora-coreos-config/pull/105 in first.

ajeddeloh commented:
I think I want to leave the latter set of commits separate. Makes it clear what changes for UEFI vs removing anaconda in general.

jlebon commented:
Testing this locally (privileged), I get:

```
+ mount /dev/sdb1 /srv/fcos/cache
mount: /srv/fcos/cache: special device /dev/sdb1 does not exist.
+ chown -h -R jlebon:jlebon /srv/fcos
chown: changing ownership of '/srv/fcos/cache/pkgcache-repo/config': Operation not permitted
chown: changing ownership of '/srv/fcos/cache/pkgcache-repo/.lock': Operation not permitted
chown: changing ownership of '/srv/fcos/cache/pkgcache-repo/tmp/cache': Operation not permitted
chown: changing ownership of '/srv/fcos/cache/pkgcache-repo/tmp': Operation not permitted
chown: changing ownership of '/srv/fcos/cache/pkgcache-repo/extensions': Operation not permitted
chown: changing ownership of '/srv/fcos/cache/pkgcache-repo/state': Operation not permitted
chown: changing ownership of '/srv/fcos/cache/pkgcache-repo/refs/heads/rpmostree/pkg/openssl-libs/1_3A1.1.1-3.fc29.x86__64': Operation not permitted
...
```

So there are two issues here. The first is https://github.com/coreos/coreos-assembler/pull/560#issuecomment-503556169. The second is that:

```sh
# ensure the user of files created do not have root ownership
trap 'chown -h -R ${USER}:${USER} ${workdir}' EXIT
```

which is now running in the privileged path for the first time. This will fail though, because the cache has a bunch of stuff owned by root and we're trying to chown it to something else as real UID != 0. I'm actually not entirely sure why we need this, since we're running supermin as the current user, so everything created through the 9p mount should already be owned by that user.

Dropped that, and got farther (it created the partitions at least), but now hitting:

```
error: Importing 90864833fa30b1aeb2ac6a5b9426bc2fdbbda227c090e70251d128d2f0835254.dirtree: fstatat(90/864833fa30b1aeb2ac6a5b9426bc2fdbbd...
```

Haven't debugged that yet. But probably comes down to the expanding ways in which one can now run cosa.

---

Tried unprivileged mode quickly. I got further but hit:

```
+ mkdir -p tmp/anaconda
+ img_base=tmp/fedora-coreos-30.1-base.qcow2
++ find /usr/lib/coreos-assembler-anaconda/ -name '*CHECKSUM'
++ head -1
+ checksum_location=/usr/lib/coreos-assembler-anaconda/Fedora-Everything-29-1.2-x86_64-CHECKSUM
+ '[' -n 1 ']'
+ img_qemu=fedora-coreos-30.1-qemu.qcow2
+ '[' x86_64 == x86_64 ']'
+ build_image
+ local size kargs
++ python3 -c 'import sys, yaml; print(yaml.load(sys.stdin)["size"])'
-c:1: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.py.
+ size=8G
++ python3 -c 'import sys, yaml; args = yaml.load(sys.stdin)["extra-kargs"]; print(" ".join(args))'
-c:1: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.py.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
KeyError: 'extra-kargs'
+ kargs=
```

Just needs a `.get("extra-args", [])` to make this work I think. Also, could use `safe_load` to squash those warnings.

jlebon commented:
OK, right also needs:

```diff
diff --git a/src/create_disk.sh b/src/create_disk.sh
index a0281a2..e1029e7 100755
--- a/src/create_disk.sh
+++ b/src/create_disk.sh
@@ -42,7 +42,7 @@ mount "${disk}2" rootfs/boot/efi

 # init the ostree
 ostree admin init-fs rootfs
-ostree pull-local "$ostree" --repo rootfs/ostree/repo
+ostree pull-local "$ostree" "$ref" --repo rootfs/ostree/repo
 ostree admin os-init "$os_name" --sysroot rootfs
 allkargs='root=/dev/disk/by-label/root rootflags=defaults,prjquota rw $ignition_firstboot'
 allkargs="$allkargs $extrakargs"
```

jlebon commented:
With the workarounds for the above, I got this working in both the privileged and unprivileged paths! :tada:

ajeddeloh commented:
Ok, just pushed the $ref and python loader warning/.get() fixes. Lets fix the other unrelated issues in their own PRs

jlebon commented:
Awesome, let's do this!

jlebon commented:
> So there are two issues here

Both fixed in https://github.com/coreos/coreos-assembler/pull/564.
